modules/iknowpy/iknowpy/engine.pyx

```diff
@@ -41,75 +43,113 @@ cdef char* aType_to_str(Attribute t) except NULL:
 	raise ValueError('Attribute type {} is unrecognized.'.format(t))
 
 
-cdef class UserDictionary:
+Entry = namedtuple("Entry", ["literal", "label"])
+
+cdef class UserDictionary(object):
 	"""A class that represents a user dictionary"""
 	cdef CPPUserDictionary user_dictionary
 
-	def __cinit__(self):
-		"""Initialize the underlying C++ iKnowUserDictioanry class"""
+	# keep track of entries in a readable form
+	cdef public entries
+
+	def __cinit__(self, load_entries=[]):
+		"""Initialize the underlying C++ iKnowUserDictionary class"""
 		self.user_dictionary = CPPUserDictionary()
 
+	def __init__(self, load_entries=[]):
+		self.entries = []
+		# load one by one so they also trigger the C++ side
+		if isinstance(load_entries, UserDictionary):
+			load_entries = load_entries.entries
+		for e in load_entries:
+			if isinstance(e, Entry):
+				self.add_label(e.literal, e.label)
+			else:
+				self.add_label(e['literal'], e['label'])
+
 	@cython.binding(True)
 	def clear(self) -> None:
 		"""Clear the User Dictionary object"""
-		return self.user_dictionary.clear()
+		self.user_dictionary.clear()
+		self.entries = []
 
 	@cython.binding(True)
 	def add_label(self, str literal: typing.Text, str UdctLabel: typing.Text) -> None:
 		"""Add a custom user dictionary label."""
-		if self.user_dictionary.addLabel(literal, UdctLabel) == 0:
-			return
+		# capture pseudo-labels for sentence end/noend
+		if (UdctLabel == Labels.SENTENCE_END):
```

Comment:
Stylistically, it is better to use if-elif-else instead of nested if-statements. Also, using parentheses in if-statements is unnecessary and discouraged.